src/librustc/ty/util.rs

```diff
@@ -1116,6 +1124,8 @@ fn needs_drop_raw<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
 
         ty::TyClosure(def_id, ref substs) => substs.upvar_tys(def_id, tcx).any(needs_drop),
 
+        ty::TyGenerator(..) => true,
```

Comment:
Generators always have a destructor which is based on the generated MIR. We don't know if this destructor is a no-op until after we run the state transformation pass.